cmd/src/code_intel_upload_transfer.go

```diff
@@ -0,0 +1,704 @@
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/sourcegraph/conc/pool"
+	"github.com/sourcegraph/sourcegraph/lib/codeintel/upload"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/sourcegraph/lib/output"
+)
+
+func UploadUncompressedIndex(ctx context.Context, filename string, httpClient upload.Client, opts upload.UploadOptions) (int, error) {
+	originalReader, originalSize, err := openFileAndGetSize(filename)
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		_ = originalReader.Close()
+	}()
+
+	bars := []output.ProgressBar{{Label: "Compressing", Max: 1.0}}
+	progress, _, cleanup := logProgress(
+		opts.Output,
+		bars,
+		"Index compressed",
+		"Failed to compress index",
+	)
+
+	compressedFile, err := compressReaderToDisk(originalReader, originalSize, progress)
+	if err != nil {
+		cleanup(err)
+		return 0, err
+	}
+	defer func() {
+		_ = os.Remove(compressedFile)
+	}()
+
+	compressedReader, compressedSize, err := openFileAndGetSize(compressedFile)
+	if err != nil {
+		cleanup(err)
+		return 0, err
+	}
+	defer func() {
+		_ = compressedReader.Close()
+	}()
+
+	cleanup(nil)
+
+	if opts.Output != nil {
+		opts.Output.WriteLine(output.Linef(
+			output.EmojiLightbulb,
+			output.StyleItalic,
+			"Indexed compressed (%.2fMB -> %.2fMB).",
+			float64(originalSize)/1000/1000,
+			float64(compressedSize)/1000/1000,
+		))
+	}
+
+	if compressedSize <= opts.MaxPayloadSizeBytes {
+		return uploadIndex(ctx, httpClient, opts, compressedReader, compressedSize, &originalSize)
+	}
+
+	return uploadMultipartIndex(ctx, httpClient, opts, compressedReader, compressedSize, &originalSize)
```

Comment:
Let's call `UploadCompressedIndex` here since this is conceptually the same? Or do you think the two code paths may need to diverge later?